(PDB-4487) Fixing Standalone Nodes using SSLContext

craigwatson opened the pull request:
As part of work to migrate to `Puppet::SSL::SSLContext`, we have unintentionally broken
compatibility with standalone/masterless nodes, who don't have a local CA or cert chain
to validate.

This commit introduces a new configuration variable (name TBC), and uses it to either
inherit from Puppet's SSL context (if we are running with a master and certificates)
or allows simple CA/CRL validation in the event that the PuppetDB server is using
either self-signed or commodity certificates.

Handoff to Puppet for these settings is via the configuration options `localcacert`
and `certificate_revocation` - if using a commodity certificate, `localcacert` should
be pointed to your OS local trust-store, and `certificate_revocation` set to `false`. If
using a self-signed certificate, this needs to be provisioned to each standalone node.

puppetlabs-jenkins commented:
Can one of the admins verify this patch?

craigwatson commented:
My knowledge of testing/documentation rules for PuppetDB are non-existent, so I'm happy to either take guidance on this and evolve my PR, or I'm equally as happy for my fork's commit to be cherry-picked out somehow and used by someone with enough skill to write adequate docs/tests

puppetcla commented:
Waiting for CLA signature by @craigwatson

@craigwatson - We require a Contributor License Agreement (CLA) for people who contribute to Puppet, but we have an easy click-through license with instructions, which is available at https://cla.puppet.com/

Note: if your contribution is trivial and you think it may be exempt from the CLA, please post a short reply to this comment with details. http://docs.puppet.com/community/trivial_patch_exemption.html

puppetcla commented:
CLA signed by all contributors.

craigwatson commented:
@joshcooper understood, thanks - I’ll get something together soon for earlier versions (also, which version should I be using as a minimum?)

I’m also considering renaming the config option to `verify_client_certificate` to be less generic, does that sound OK?

craigwatson commented:
@austb @joshcooper I've now added backwards-compatible code to handle Puppet < `6.4.0` and fixed the tests accordingly. I'm not _entirely_ happy with the solution as I'm not a huge fan of if-version statements, but it feels the cleanest way to do it.

I've also renamed the config parameter to `verify_client_certificate`.

Feel free to let me know if there's anything else to change - the spec tests pass on both Puppet 6.3.0 and 6.4.0 up to 6.8.0 :)

craigwatson commented:
Thanks @joshcooper and @austb for the approvals 🙂

Would you need the standalone documentation edited within this PR?

austb commented:
@craigwatson I don't mind either way, if you'd like to take a stab at it (just let me know), otherwise I'll go ahead and merge this PR and edit the documentation separately.

craigwatson commented:
@austb no worries, feel free to merge as-is.

Is there a timeline for the `puppetdb-termimi` package release?

austb commented:
@craigwatson I believe we are targeting a release for September 17. But it can be delayed a few days if there are issues during release.

lofic commented:
Hello, I've tested with puppetdb-termini-6.6.0-1.el7.noarch released yesterday and it's not better.


With puppetdb-termini-6.6.0 :

```
# /opt/puppetlabs/puppet/bin/puppet apply   --modulepath=/deploy/environments/dev/modules   -e 'include base'

Error: Failed to initialize SSL: The CA certificates are missing from '/etc/puppetlabs/puppet/ssl/certs/ca.pem'
Error: Run `puppet agent -t`
Error: Cached facts for centos7-sandbox-acl.c.int-lz1-dtp.internal failed: Failed to find facts from PuppetDB at puppet:8140: The CA certificates are missing from '/etc/puppetlabs/puppet/ssl/certs/ca.pem'
Error: Failed to initialize SSL: The CA certificates are missing from '/etc/puppetlabs/puppet/ssl/certs/ca.pem'
Error: Run `puppet agent -t`
Error: The CA certificates are missing from '/etc/puppetlabs/puppet/ssl/certs/ca.pem'
Error: Could not run: Could not retrieve facts for centos7-sandbox-acl.c.int-lz1-dtp.internal: The CA certificates are missing from '/etc/puppetlabs/puppet/ssl/certs/ca.pem'
Wrapped exception:
The CA certificates are missing from '/etc/puppetlabs/puppet/ssl/certs/ca.pem
```

With puppetdb-termini-6.4.0 (after 2 `yum downgrade puppetdb-termini`) :

```
# /opt/puppetlabs/puppet/bin/puppet apply   --modulepath=/deploy/environments/dev/modules   -e 'include base'
Warning: Puppet::SSL::Host is deprecated and will be removed in a future release of Puppet.
   (location: /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/ssl/host.rb:235:in `initialize')
Error: Failed to open TCP connection to puppet:8140 (getaddrinfo: Name or service not known)
Error: Cached facts for centos7-sandbox-acl.c.int-lz1-dtp.internal failed: Failed to find facts from PuppetDB at puppet:8140: Failed to open TCP connection to puppet:8140 (getaddrinfo: Name or service not known)
Notice: Compiled catalog for centos7-sandbox-acl.c.int-lz1-dtp.internal in environment dev in 0.30 seconds
Notice: Applied catalog in 1.60 seconds
```

The apply is OK although I still have a non blocking error related (I guess) to another issue (https://tickets.puppetlabs.com/browse/PUP-9314 : puppetdb-termini for puppet standalone try to connect to master)

craigwatson commented:
@lofic As mentioned on the original JIRA ticket (just to make sure that people who see this GitHub PR see the same resolution), you'll need to edit both `puppet.conf` and `puppetdb.conf`:

`puppet.conf`:
```
verify_client_certificate = false
```

`puppetdb.conf`:
```
  localcacert            = /etc/pki/tls/certs/ca-bundle.crt
  certificate_revocation = false
```

lofic commented:
Hello, thanks, but it's not better unfortunately : 

```#  rpm -qa | grep puppet
puppet-agent-6.9.0-1.el7.x86_64
puppetdb-termini-6.6.0-1.el7.noarch
puppet6-release-6.0.0-5.el7.noarch

# cat /etc/puppetlabs/puppet/puppet.conf 
[main]

report=true
reports=puppetdb

storeconfigs=true
storeconfigs_backend=puppetdb

environmentpath=/deploy/environments

verify_client_certificate = false

# cat /etc/puppetlabs/puppet/puppetdb.conf 
[main]
server_urls = https://puppetdb:8081
soft_write_failure = false
certificate_revocation = false
localcacert            = /etc/pki/tls/certs/ca-bundle.crt

# cat /etc/puppetlabs/puppet/routes.yaml 
---
apply:
  catalog:
    terminus: compiler
    cache: puppetdb
  resource:
    terminus: ral
    cache: puppetdb
  facts:
    terminus: facter
    #cache: puppetdb_apply

# /opt/puppetlabs/puppet/bin/puppet apply --modulepath=/deploy/environments/dev/module -e "package {'mutt': ensure => present }"

Error: Failed to initialize SSL: The CA certificates are missing from '/etc/puppetlabs/puppet/ssl/certs/ca.pem'
Error: Run `puppet agent -t`
Error: Cached facts for centos7-sandbox-acl.c.int-lz1-dtp.internal failed: Failed to find facts from PuppetDB at puppet:8140: The CA certificates are missing from '/etc/puppetlabs/puppet/ssl/certs/ca.pem'
Error: Failed to initialize SSL: The CA certificates are missing from '/etc/puppetlabs/puppet/ssl/certs/ca.pem'
Error: Run `puppet agent -t`
Error: The CA certificates are missing from '/etc/puppetlabs/puppet/ssl/certs/ca.pem'
Error: Could not run: Could not retrieve facts for centos7-sandbox-acl.c.int-lz1-dtp.internal: The CA certificates are missing from '/etc/puppetlabs/puppet/ssl/certs/ca.pem'
Wrapped exception:
The CA certificates are missing from '/etc/puppetlabs/puppet/ssl/certs/ca.pem'

# yum -y downgrade puppetdb-termini &>/dev/null
# yum -y downgrade puppetdb-termini &>/dev/null
#  rpm -qa | grep puppet
puppet-agent-6.9.0-1.el7.x86_64
puppetdb-termini-6.4.0-1.el7.noarch
puppet6-release-6.0.0-5.el7.noarch

# /opt/puppetlabs/puppet/bin/puppet apply --modulepath=/deploy/environments/dev/module -e "package {'mutt': ensure => present }"

Warning: Puppet::SSL::Host is deprecated and will be removed in a future release of Puppet.
   (location: /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/ssl/host.rb:235:in `initialize')
Error: Failed to open TCP connection to puppet:8140 (getaddrinfo: Name or service not known)
Error: Cached facts for centos7-sandbox-acl.c.int-lz1-dtp.internal failed: Failed to find facts from PuppetDB at puppet:8140: Failed to open TCP connection to puppet:8140 (getaddrinfo: Name or service not known)
Notice: Compiled catalog for centos7-sandbox-acl.c.int-lz1-dtp.internal in environment production in 0.69 seconds
Notice: /Stage[main]/Main/Package[mutt]/ensure: created
Notice: Applied catalog in 1.96 seconds
```

lofic commented:
I confirm that it works like a charm, and it also fixed my other problem !

Provided that the puppetdb SSL proxy cert is signed by a CA trusted in /etc/pki/tls/certs/ca-bundle.crt

Because I have a custom SSL cert and CA for the SSL proxy of the puppetdb, I had to add it (with update-ca-trust on CentOS)

All good.

Thanks a lot !

I agree with you that this information should be added to the doc here :

https://puppet.com/docs/puppetdb/6.6/connect_puppet_apply.html#option-a-set-up-an-ssl-proxy-for-puppetdb

craigwatson commented:
@lofic Ah, I should have also explained this!

So the `localcacert` configuration option should point to the location of your CA certificate - my particular instance of PuppetDB is fronted by a load balancer, which uses a certificate signed by Lets Encrypt, which is _itself_ distributed as part of the OS' native trust bundle.

If you're using your own CA, you can simply point `localcacert` to your CA on disk.

I'll get a PR in at some point to fix up the documentation :)
